gson/src/main/java/com/google/gson/GsonBuilder.java

```diff
@@ -718,14 +728,16 @@ public GsonBuilder registerTypeAdapterFactory(TypeAdapterFactory factory) {
    * @param typeAdapter This object must implement at least one of {@link TypeAdapter},
    *        {@link JsonSerializer} or {@link JsonDeserializer} interfaces.
    * @return a reference to this {@code GsonBuilder} object to fulfill the "Builder" pattern
+   * @throws IllegalArgumentException if the Type adapter being registered is for a JsonElements or any of its subclasses
    * @since 1.7
    */
   @CanIgnoreReturnValue
   public GsonBuilder registerTypeHierarchyAdapter(Class<?> baseType, Object typeAdapter) {
     Objects.requireNonNull(baseType);
-    $Gson$Preconditions.checkArgument(typeAdapter instanceof JsonSerializer<?>
+    $Gson$Preconditions.checkArgument((typeAdapter instanceof JsonSerializer<?>
         || typeAdapter instanceof JsonDeserializer<?>
-        || typeAdapter instanceof TypeAdapter<?>);
+        || typeAdapter instanceof TypeAdapter<?>)
+        && !JsonElement.class.isAssignableFrom(baseType));
```

Comment:
I think it would be better to have this check as separate statement, and then also use a more useful exception message, e.g.:
```java
if (JsonElement.class.isAssignableFrom(baseType)) {
  throw new IllegalArgumentException("Cannot override built-in adapter for " + baseType);
}
```
(or similar)